nautobot/extras/models/approvals.py

```diff
@@ -0,0 +1,228 @@
+from django.contrib.auth.models import Group
+from django.contrib.contenttypes.fields import GenericForeignKey
+from django.contrib.contenttypes.models import ContentType
+from django.db import models
+
+from nautobot.core.constants import CHARFIELD_MAX_LENGTH
+from nautobot.core.models.generics import PrimaryModel
+from nautobot.extras.choices import ApprovalWorkflowStateChoices
+from nautobot.extras.constants import APPROVAL_WORKFLOW_MODELS
+from nautobot.extras.models import ChangeLoggedModel
+from nautobot.extras.utils import extras_features
+from nautobot.users.models import User
+
+
+@extras_features(
+    "custom_links",
+    "custom_validators",
+    "export_templates",
+    "graphql",
+    "relationships",
+    "webhooks",
+)
+class ApprovalWorkflow(PrimaryModel):
+    """ApprovalWorkflow model."""
+
+    name = models.CharField(
+        max_length=CHARFIELD_MAX_LENGTH,
+        unique=True,
+    )
+    model_content_type = models.ForeignKey(
+        to=ContentType,
+        on_delete=models.PROTECT,
+        related_name="+",
+        limit_choices_to=APPROVAL_WORKFLOW_MODELS,
+    )
+    model_constraints = models.JSONField(
+        blank=True,
+        null=True,
+    )
+
+    class Meta:
+        """Meta class for ApprovalWorkflow."""
+
+        verbose_name = "Approval Workflow"
+
+    def __str__(self):
+        """Stringify instance."""
+        return self.name
+
+
+@extras_features(
+    "custom_links",
+    "custom_validators",
+    "export_templates",
+    "graphql",
+    "webhooks",
+)
+class ApprovalWorkflowStage(PrimaryModel):
+    """ApprovalWorkflowStage model."""
+
+    approval_workflow = models.ForeignKey(
+        to="extras.ApprovalWorkflow",
+        related_name="approval_workflow_stages",
+        verbose_name="Approval Workflow",
+        on_delete=models.PROTECT,
+        help_text="Approval workflow to which this stage belongs.",
+    )
+    sequence_weight = models.PositiveIntegerField(
+        help_text="The weight dictates the order in which this stage will need to be approved. The lower the number, the earlier it will be.",
+    )
+    name = models.CharField(max_length=CHARFIELD_MAX_LENGTH)
+    min_approvers = models.PositiveIntegerField(
+        help_text="Number of minimum approvers required to approve this stage.",
+    )
+    denial_message = models.CharField(
+        max_length=CHARFIELD_MAX_LENGTH, blank=True, help_text="Message to show when the stage is denied."
+    )
+    approver_group = models.ForeignKey(
+        to=Group,
+        related_name="approval_workflow_stages",
+        verbose_name="Group",
+        help_text="Group of users who are eligible to approve this stage.",
+        on_delete=models.PROTECT,
+    )
+
+    class Meta:
+        """Meta class for ApprovalWorkflowStage."""
+
+        verbose_name = "Approval Workflow Stage"
+        unique_together = [["approval_workflow", "name"], ["approval_workflow", "sequence_weight"]]
```

Comment:
Should we allow stages with the same `sequence_weight` what is that gonna look like?

Reply:
Yeah so `sequence_weight` determines the order in which the approval workflow stage should be approved. So if two or more stages have the same weight, which stage should be sent to the user first for approval essentially. The more I think about it, the more I think the current uniqueness constraint makes sense.

Reply:
I agree with you here. `sequence_weight` sounds a little confusing to me as well. I think the underlying motivation for us to name this field `sequence_weight` is that we have been using `weight` to order stuff (nav menu items, UI components and etc.) in Nautobot so that the users are already familiar with how it works on a high level. What do you guys think? @glennmatthews @lampwins

Reply:
I think I was the one voting in favor of just `sequence` originally, but perhaps we should just go with `weight` for consistency with things like Dynamic Groups and Config Contexts.

To Daria's comment, `weight` doesn't need to be globally unique, but it does need to be unique within a given workflow. That's accomplished here by the `unique_together = [...["approval_workflow", "sequence_weight"]]` portion.